src/aggregates/vector_selector.rs

```diff
@@ -0,0 +1,404 @@
+use pgx::*;
+
+use pgx::error;
+
+use crate::aggregate_utils::in_aggregate_context;
+use crate::aggregates::{Milliseconds, STALE_NAN, USECS_PER_MS};
+use serde::{Deserialize, Serialize};
+
+use crate::palloc::{Inner, InternalAsValue, ToInternal};
+use crate::raw::{bytea, TimestampTz};
+
+// a vector selector aggregate has the same semantics as parse.VectorSelector processing
+// in Prometheus. Namely, for all timestamps ts in the series:
+//     ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval
+// we return the last sample.value such that the time sample.time <= ts and sample.time >= ts-lookback.
+// if such a sample doesn't exist return NULL (None).
+// thus, the vector selector returns a regular series of values corresponding to all the points in the
+// ts series above.
+// Note that for performance, this aggregate is parallel-izable, combinable, and does not expect ordered inputs.
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_transition(
+    state: Internal,
+    start_time: TimestampTz,
+    end_time: TimestampTz,
+    bucket_width: Milliseconds,
+    lookback: Milliseconds,
+    time: TimestampTz,
+    value: f64,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Internal {
+    vector_selector_transition_inner(
+        unsafe { state.to_inner() },
+        start_time.into(),
+        end_time.into(),
+        bucket_width,
+        lookback,
+        time.into(),
+        value,
+        fcinfo,
+    )
+    .internal()
+}
+pub fn vector_selector_transition_inner(
+    state: Option<Inner<VectorSelector>>,
+    start_time: pg_sys::TimestampTz,
+    end_time: pg_sys::TimestampTz,
+    bucket_width: Milliseconds,
+    lookback: Milliseconds,
+    time: pg_sys::TimestampTz,
+    value: f64,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<VectorSelector>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let mut state = state.unwrap_or_else(|| {
+                let state: Inner<VectorSelector> =
+                    VectorSelector::new(start_time, end_time, bucket_width, lookback).into();
+                state
+            });
+
+            state.insert(time, value);
+
+            Some(state)
+        })
+    }
+}
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_final(
+    state: Internal, /* Option<Inner<VectorSelector>> */
+) -> Option<Vec<Option<f64>>> {
+    let state: Option<Inner<VectorSelector>> = unsafe { state.to_inner() };
+    state.map(|s| s.to_pg_array())
+}
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_serialize(state: Internal) -> bytea {
```

Comment:
is this ever hit in tests?

Reply:
I _think_ (based on my reading of the PG docs on this topic https://www.postgresql.org/docs/14/xaggr.html#XAGGR-PARTIAL-AGGREGATES) that now that we have a parallel execution test, this function should now also be hit.

Reply:
Based on the coverage report that I was able to generate, these functions are not being hit by the tests. I'm having some issues with debugging this at the moment though. Will keep digging.

Reply:
Have you tried breaking the logic and seeing if tests fail? It may be a misreport by the coverage tool

Reply:
The coverage was correct. I managed to figure out how to force PG into performing a partial aggregation, and it surfaced a bug. Partial aggregation on `vector_select` is broken (also in previous releases - I tested `0.2.0` and `0.3.0` and has probably never worked. That might be interesting in itself, because obviously we're not generating queries in `promscale` that _would_ use it.

Reply:
And I found out how to fix it.